File: java/src/main/java/com/eronalves/functionalendpoints/user/Fetcher.java
package com.eronalves.functionalendpoints.user;


import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class Fetcher {

    @Autowired
    MongoRepo repository;

    public List<User> all(){
        return repository.findAll();
    }

    public User insert(User user) {
        return repository.save(user);
    }

    public User get(String id) {
        return findUser(id);
    }

    public void update(User received, String id) {
        User user = findUser(id);
        if(!received.id().equals(id)) throw new RuntimeException("The ID's are not equal");
        user.updateFrom(received);
    }

    public void delete(String id) {
        repository.deleteById(id);
    }

    private User findUser(String id){
        Optional<User> possibleUser = repository.findById(id);
        if(possibleUser.isEmpty()) throw new RuntimeException("User not found");
        return possibleUser.get();
    }

}

File: kotlin/src/main/kotlin/com/eronalves/functionalendpoints/user/Fetcher.kt
package com.eronalves.functionalendpoints.user

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class Fetcher(private val repository: UserRepo) {
    fun all(): List<User> {
        return repository.findAll()
    }

    fun insert(user: User): User {
        return repository.save(user)
    }

    fun getById(id: String): User {
        return findUser(id)
    }

    @Transactional
    fun update(user: User, id: String) {
        if(user.id != id) throw RuntimeException("The ID's are not equal")
        val userFinded = findUser(id)
        userFinded.updateFrom(user)
    }

    fun delete(id: String) {
        repository.deleteById(id)
    }

    private fun findUser(id: String): User {
        val possibleUser = repository.findById(id)
        if (possibleUser.isEmpty) throw RuntimeException("User not found")
        return possibleUser.get()
    }

}
